stan/math/rev/core/var.hpp

```diff
@@ -25,36 +23,49 @@ static void grad(vari* vi);
  * an arena-based memory manager scoped to a single gradient
  * calculation.
  *
- * A var is constructed with a double and used like any
+ * A var is constructed with a type `T` and used like any
  * other scalar.  Arithmetical functions like negation, addition,
  * and subtraction, as well as a range of mathematical functions
  * like exponentiation and powers are overridden to operate on
  * var values objects.
+ * @tparam T An Arithmetic type.
  */
-class var {
+template <typename T, typename = void>
+class var_value {};
+
+template <typename T>
+class var_value<T, require_floating_point_t<T>> {
```

Comment:
If we don't do the `S -> T` conversion, does the second template argument go away?

Reply:
I'm not sure I'm following, but the second template function here just forces `T` to be a floating point type. I also don't like this second template here but the only way I found to get rid of them is kind of nasty imo and involve an underlying `Nested` class in the `var` class

Reply:
Got it.